### PR TITLE
Metadata parser: add support for built-in authentication profiles and metadata properties

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
Completes some tasks that I've had in my TODO for a while with regards to the metadata parser.

## Built-in authentication profiles

Adds built-in authentication profiles, which will make it much easier to author (and maintain) metadata.yaml files for Azure (and in the future other environments), reducing boilerplate code.

Now, if a component supports Azure AD for authentication, it can be defined in the metadata as:

```yaml
builtinAuthenticationProfiles:
  - name: "azuread"
```

If the authentication profile requires additional properties, they can be included too:

```yaml
builtinAuthenticationProfiles:
  - name: "azuread"
    metadata:
      - name: namespaceName
        description: "Parameter to set the address of the Service Bus namespace, as a fully-qualified domain name."
        example: '"namespace.servicebus.windows.net"'
        required: true
        binding:
          output: true
          input: true
```

The parser will automatically populate the 3 Azure AD authentication methods.

This is in addition to `authenticationProfiles`, which continues to work as before.

## Built-in metadata properties

Certain components have "built-in" metadata properties that are parsed by the runtime so they're normally not defined in the components' metadata structs.

These properties are now added automatically and should not be included in the components' metadata.yaml.

- State stores:
  - `actorStateStore` (which will be added only if the component has the capability "actorStateStore")
  - `keyPrefix`
- Locks: 
  - `keyPrefix`
- PubSub:
  - `consumerID`

### CODE REVIEW

**Comment by berndverst** [comment]
Nice. Merge after my PubSub PR please. I will withhold my approval until that time to ensure the merge order is maintained.

File: .build-tools/pkg/metadataanalyzer/utils.go
Comment (berndverst) [inline .build-tools/pkg/metadataanalyzer/utils.go:84]: Please note these are the folder names here and not necessarily the component names!!

They are intentionally distinct from what you found in mdutils because the component category folder name might not match the component type exactly.

**Comment by ItalyPaleAle** [comment]
@berndverst your pubsub PR is merged already.

Thanks for the review, I'll address the comment tomorrow

**Comment by berndverst** [comment]
> @berndverst your pubsub PR is merged already.
> 
> 
> 
> Thanks for the review, I'll address the comment tomorrow

Oh thanks! Didn't notice.

This PR is awesome by the way!

**Comment by ItalyPaleAle** [comment]
@berndverst thanks!

Reverted the change, should be good now

**Comment by berndverst** [comment]
@ItalyPaleAle Good thinking but I'm concerned that this will confuse folks. `consumerID` is used slightly differently in different components - for that reason I would still suggest a manual description for consumerID so this can be explained.
Also at least one component (I think it was jetstream) does not use consumerID right now.

As for `actorStateStore` - have we actually added this capability in every supported state store today? I fear it might be missing somewhere.

And is `keyPrefix` truly supported in state store? Is that handled by runtime?

**Comment by ItalyPaleAle** [comment]
> consumerID is used slightly differently in different components - for that reason I would still suggest a manual description for consumerID so this can be explained

I'm not sure about that so I'll defer to you

> As for actorStateStore - have we actually added this capability in every supported state store today? I fear it might be missing somewhere.

This is implemented by adding `actorStateStore` in the metadata only if the metadata.yaml includes the capability `actorStateStore`.

> And is keyPrefix truly supported in state store? Is that handled by runtime?

Yes on both here. Also supported in every lock store.

**Comment by ItalyPaleAle** [comment]
@berndverst I have removed "consumerID" as a built-in property given your comments above

**Comment by ItalyPaleAle** [comment]
Given the experience with Kafka yesterday, I would actually lean towards no. It really seems that every component does what it wants :( How about we park this one and review at a later time?

**Comment by berndverst** [comment]
Want to remove the consumerID stuff then for now (keep the code around somewhere else just in case) so we can merge the rest?

**Review by berndverst — APPROVED** [review]

File: metadata/utils.go
Comment (ItalyPaleAle) [reply metadata/utils.go:256]: I needed those to match the name of the component type in the component spec, such as `bindings.foo`

**Pull request merged**
